### PR TITLE
GEODE-3288: Replaces char* with std::string on public APIs.

### PR DESCRIPTION
**Pull request by pivotal-jbarrett**
- Replaces char/wchar_t arrays with char16_t arrays.
- Replaces char*/wchar_t* strings arrays with std::vector<std::string>.
- Fixes type for reading byte arrays.
- Replaces unsigned and signed char * for byte arrays with int8_t*.
- Remove char* constructor from CacheableString
- Converts Serializer to use std::string.
- Adds methods for std::string consistent with Java DataOutput.
- Removes writeASCII from public API.
- Adds JavaModifiedUtf8 helper class.
- Removes DataOuput::getEncodedLength from public API.

### CODE REVIEW

File: tests/cpp/testobject/PdxClassV1.cpp [hunk 1]
```diff
@@ -560,69 +550,69 @@ std::string PdxTypesIgnoreUnreadFieldsV1::toString() const {
...
+  m_floatArray = new float[size];
+  m_doubleArray = new double[size];
+
+  for (int i = 0; i < size; i++) {
+    m_int16Array[i] = 0x2332;
+    m_int32Array[i] = 34343 + i;
+    m_longArray[i] = 324324L + i;
+    m_floatArray[i] = 232.565f + i;
+    m_doubleArray[i] = 23423432 + i;
+    // m_stringArray[i] = String.Format("one{0}", i);
```
Comment (mhansonp) [inline tests/cpp/testobject/PdxClassV1.cpp]: This was commented out in the original code, but thought I'd point it out.  Not used?  Not supported?

File: cppcache/integration-test/BuiltinCacheableWrappers.hpp [hunk 2]
```diff
@@ -349,16 +352,14 @@ class CacheableFileNameWrapper : public CacheableWrapper {
...
   virtual uint32_t getCheckSum(const std::shared_ptr<Cacheable> object) const {
-    const CacheableFileName* obj =
-        dynamic_cast<const CacheableFileName*>(object.get());
-    return (obj != nullptr ? CacheableHelper::crc32(
-                                 (uint8_t*)obj->value().c_str(), obj->length())
-                           : 0);
+    auto&& obj = std::dynamic_pointer_cast<CacheableFileName>(object);
+    return (obj ? CacheableHelper::crc32((uint8_t*)obj->value().c_str(),
+                                         obj->length())
+                : 0);
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/integration-test/BuiltinCacheableWrappers.hpp]: It is consistent with the style guide. The `:` is aligned with the `?`.

File: cppcache/integration-test/testThinClientCqDelta.cpp [hunk 3]
```diff
@@ -56,7 +56,8 @@ class CqDeltaListener : public CqListener {
     auto deltaValue = aCqEvent.getDeltaValue();
     DeltaTestImpl newValue;
     auto input = getHelper()->getCache()->createDataInput(
-        deltaValue->value(), deltaValue->length());
+        reinterpret_cast<const uint8_t*>(deltaValue->value()),
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/integration-test/testThinClientCqDelta.cpp]: DataInput/Output work from C++ byte buffers, in our case that is `uint8_t*` (for better or worse). The value from the delta is Geode (Java) byte array or `int8_t*`. A cast is required.

File: cppcache/src/util/JavaModifiedUtf8.hpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      return 0;
+    }
+
+    // TODO string optimize for direct calculation
+    return encodedLength(to_utf16(utf8));
+  }
+
+  /**
+   * Calculate the length of the given UTF-16 string when encoded in Java
+   * Modified UTF-8.
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/JavaModifiedUtf8.hpp]: No.

File: cppcache/src/util/JavaModifiedUtf8.hpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  inline static size_t encodedLength(const std::u16string& utf16) {
+    return encodedLength(utf16.data(), utf16.length());
+  }
+
+  inline static size_t encodedLength(const char16_t* data, size_t length) {
+    size_t encodedLen = 0;
+    while (length-- > 0) {
+      const char16_t c = *(data++);
+      if (c == 0) {
+        // NUL
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/JavaModifiedUtf8.hpp]: `NUL` is another standard abbreviation of the null character and the comment was copied from the original source.

File: cppcache/src/util/JavaModifiedUtf8.hpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        encodedLen += 2;
+      } else {
+        encodedLen += 3;
+      }
+    }
+    return encodedLen;
+  }
+
+  /**
+   * Converts given UTF-8 string to Java Modified UTF-8 string.
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/JavaModifiedUtf8.hpp]: Converts from java modified UTF-8 to to UTF-8. The implementation does so by passing through UTF-16 first.

File: cppcache/integration-test/testThinClientBigValue.cpp [hunk 7]
```diff
@@ -71,16 +71,17 @@ void verify(std::shared_ptr<CacheableBytes>& valuePtr, int size) {
     SLEEP(100);
   }
 
+  printf("check size[%d]\n", size);
   ASSERT(valSize == size, msg);
 
+  auto&& bytes = reinterpret_cast<const uint8_t*>(valuePtr->value());
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/integration-test/testThinClientBigValue.cpp]: No. The original test expected `uint8_t*` so the path of least resistance keeps it that way.

File: cppcache/src/TcrMessage.cpp [hunk 8]
```diff
@@ -664,30 +664,17 @@ void TcrMessage::writeRegionPart(const std::string& regionName) {
 }
 
 void TcrMessage::writeStringPart(const std::string& str) {
-  const char* value = str.c_str();
-  uint32_t length = 0;
-  if (value != nullptr) {
-    int32_t encodedLen = m_request->getEncodedLength(value, length);
+  // TODO string does not correctly encode java modified utf-8 from utf-8
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/TcrMessage.cpp]: Actually this is no longer the case. Will remove the statement.

File: cppcache/src/ThinClientPoolDM.cpp [hunk 9]
```diff
@@ -696,57 +696,43 @@ GfErrType ThinClientPoolDM::sendRequestToAllServers(
...
     const {
-  int32_t size = static_cast<int32_t>(m_attrs->m_initLocList.size());
-  std::shared_ptr<CacheableString>* ptrArr =
-      new std::shared_ptr<CacheableString>[size];
-
-  for (int i = 0; i < size; ++i) {
-    ptrArr[i] = CacheableString::create(
-        m_attrs->m_initLocList[i].c_str(),
-        static_cast<int32_t>(m_attrs->m_initLocList[i].length()));
+  // TODO refactor - why return CacheableStringArray
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/ThinClientPoolDM.cpp]: https://issues.apache.org/jira/browse/GEODE-4258

File: cppcache/src/ThinClientPoolDM.cpp [hunk 10]
```diff
@@ -696,57 +696,43 @@ GfErrType ThinClientPoolDM::sendRequestToAllServers(
...
+        new std::shared_ptr<CacheableString>[m_attrs->m_initServList.size()];
+    int32_t i = 0;
+    for (auto&& server : m_attrs->m_initServList) {
+      ptrArr[i++] = CacheableString::create(server);
+    }
+    return CacheableStringArray::createNoCopy(ptrArr, i);
+  } else if (!m_attrs->m_initLocList.empty()) {
     std::vector<ServerLocation> vec;
-    ((ThinClientLocatorHelper*)m_locHelper)
+    // TODO thread - why is this member volatile?
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/ThinClientPoolDM.cpp]: https://issues.apache.org/jira/browse/GEODE-4258

File: cppcache/src/ThinClientPoolDM.cpp [hunk 11]
```diff
@@ -696,57 +696,43 @@ GfErrType ThinClientPoolDM::sendRequestToAllServers(
...
-    }
-  }
-
-  if (size > 0) return CacheableStringArray::createNoCopy(ptrArr, size);
-
-  size = static_cast<int32_t>(m_attrs->m_initLocList.size());
-
-  // get dynamic added servers using locators
-  if (size > 0) {
+  // TODO refactor - why return CacheableStringArray
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/ThinClientPoolDM.cpp]: https://issues.apache.org/jira/browse/GEODE-4258

File: cppcache/src/util/JavaModifiedUtf8.hpp [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+  /**
+   * Calculate the length of the given UTF-8 string when encoded in Java
+   * Modified UTF-8.
+   */
+  inline static size_t encodedLength(const std::string& utf8) {
+    if (utf8.empty()) {
+      return 0;
+    }
+
+    // TODO string optimize for direct calculation
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/JavaModifiedUtf8.hpp]: https://issues.apache.org/jira/browse/GEODE-4189

File: cppcache/src/util/string.hpp [hunk 13]
```diff
@@ -24,10 +24,19 @@
...
 
+/**
+ * Native codecvt_mode endianess
+ */
+constexpr std::codecvt_mode codecvt_mode_native_endian =
+    endian::native == endian::little ? std::little_endian
+                                     : (std::codecvt_mode)0;
+
 inline std::u16string to_utf16(const std::string& utf8) {
 // TODO string replace with boost
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/string.hpp]: https://issues.apache.org/jira/browse/GEODE-4260

File: cppcache/src/util/string.hpp [hunk 14]
```diff
@@ -46,6 +55,80 @@ inline std::u16string to_utf16(const std::string& utf8) {
 #endif
 }
 
+inline std::u16string to_utf16(const std::u32string& ucs4) {
+// TODO string replace with boost
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/string.hpp]: https://issues.apache.org/jira/browse/GEODE-4260

File: cppcache/src/util/string.hpp [hunk 15]
```diff
@@ -46,6 +55,80 @@ inline std::u16string to_utf16(const std::string& utf8) {
...
+          char32_t>{}
+          .to_bytes(ucs4);
+#endif
+
+  return std::u16string(reinterpret_cast<const char16_t*>(bytes.c_str()),
+                        bytes.length() / sizeof(char16_t));
+}
+
+inline std::u16string to_utf16(const char32_t* ucs4, size_t len) {
+// TODO string replace with boost
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/string.hpp]: https://issues.apache.org/jira/browse/GEODE-4260

File: cppcache/src/util/string.hpp [hunk 16]
```diff
@@ -46,6 +55,80 @@ inline std::u16string to_utf16(const std::string& utf8) {
...
+          char32_t>{}
+          .to_bytes(ucs4, ucs4 + len);
+#endif
+
+  return std::u16string(reinterpret_cast<const char16_t*>(bytes.c_str()),
+                        bytes.length() / sizeof(char16_t));
+}
+
+inline std::u32string to_ucs4(const std::u16string& utf16) {
+// TODO string replace with boost
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/util/string.hpp]: https://issues.apache.org/jira/browse/GEODE-4260

**Pull request merged**